### PR TITLE
♻️ Feature flag refactor: no more "prop drilldown", now just import

### PR DESCRIPTION
**Pull request by jagracey**
## Motivation

Related to (internal) [RUMF-1004](https://datadoghq.atlassian.net/browse/RUMF-1004).

Currently prop drill-down is required to pass the configuration object around to access `configuration.isEnabled()`. This is a lot of work, and requires care to not introduce further bugs. Instead we can leverage a singleton to import `isEnabledExperimentalFeatures`.

## Changes

Moving from `configuration.isEnabled` prop drill-down  to instead `import {isEnabledExperimentalFeatures} from '@datadog/browser-core'`;

## Testing

---

I have gone over the [contributing](https://github.com/DataDog/browser-sdk/blob/main/CONTRIBUTING.md) documentation.

### CODE REVIEW

**Review by acorretti — APPROVED** [review]

File: eslint-local-rules/disallowSideEffects.js [hunk 1]
```diff
@@ -34,7 +34,7 @@ const pathsWithSideEffect = new Set([
   `${packagesRoot}/rum-slim/src/index.ts`,
 ])
 
-// Those packages are known to have no side effects when evaluated
+// Those packages either are known to have no side effects when evaluated
```
Comment (BenoitZugmeyer) [inline eslint-local-rules/disallowSideEffects.js:37]: Could you revert this change?

File: packages/core/src/boot/init.ts [hunk 2]
```diff
@@ -51,6 +52,12 @@ export function commonInit(initConfiguration: InitConfiguration, buildEnv: Build
   const configuration = buildConfiguration(initConfiguration, buildEnv)
   const internalMonitoring = startInternalMonitoring(configuration)
 
+  const enableExperimentalFeatures = Array.isArray(initConfiguration.enableExperimentalFeatures)
+    ? initConfiguration.enableExperimentalFeatures
+    : []
+
+  setEnabledExperimentalFeatures(enableExperimentalFeatures)
+
```
Comment (BenoitZugmeyer) [inline packages/core/src/boot/init.ts:60]: I think it would be better to move this at the beginning of `commonInit`, so we can use feature flags while building the configuration.

File: packages/core/src/boot/init.ts [hunk 3]
```diff
@@ -48,6 +49,11 @@ export interface BuildEnv {
 }
 
 export function commonInit(initConfiguration: InitConfiguration, buildEnv: BuildEnv) {
+  const enableExperimentalFeatures = Array.isArray(initConfiguration.enableExperimentalFeatures)
+    ? initConfiguration.enableExperimentalFeatures
+    : []
+  setEnabledExperimentalFeatures(enableExperimentalFeatures)
+
   const configuration = buildConfiguration(initConfiguration, buildEnv)
   const internalMonitoring = startInternalMonitoring(configuration)
```
Comment (bcaudan) [inline packages/core/src/boot/init.ts:58]: it could be nice to extract experimental features init to its own function to keep the same level of abstraction inside commonInit

**Review by bcaudan — COMMENTED** [review]
note on PR form:
- it would deserve a better title to clearly understand what this refactor is about
- it could be interesting to clearly explicit the design change in the changes section (`configuration.isEnabled` -> `isEnabledExperimentalFeatures`), it would be easier for reviewer to have a general idea of the change
- it seems that there was an issue when you have updated your branch with main since some changes already in main appear in your diff

File: packages/core/src/domain/configuration/experimentalFeatures.ts [hunk 4]
```diff
@@ -0,0 +1,9 @@
+let enabledExperimentalFeatures: Set<string>
+
+export function setEnabledExperimentalFeatures(enabledFeatures: string[]): void {
+  enabledExperimentalFeatures = new Set(enabledFeatures)
+}
+
+export function isEnabledExperimentalFeature(featureName: string): boolean {
+  return enabledExperimentalFeatures.has(featureName)
+}
```
Comment (bcaudan) [inline packages/core/src/domain/configuration/experimentalFeatures.ts:9]: any thoughts on usage during tests?

Comment (amortemousque) [inline packages/core/src/domain/configuration/experimentalFeatures.ts]: I have some thoughts about the naming. What do you think about:

- enableExperimentalFeatures(enabledFeatures)
- isExperimentalFeatureEnabled(featureName)

Comment (jagracey) [reply packages/core/src/domain/configuration/experimentalFeatures.ts]: This convention of {specifier + group_name} is less correct english, but it isn't uncommon on codebases.
I don't mind one way or the other. Do we have a convention of the reverse?

Comment (jagracey) [reply packages/core/src/domain/configuration/experimentalFeatures.ts:13]: Yes, a reset function makes sense. Just added `resetExperimentalFeatures`.

File: packages/core/src/boot/init.ts [hunk 5]
```diff
@@ -48,6 +49,7 @@ export interface BuildEnv {
 }
 
 export function commonInit(initConfiguration: InitConfiguration, buildEnv: BuildEnv) {
+  updateExperimentalFeatures(initConfiguration.enableExperimentalFeatures || [])
```
Comment (bcaudan) [inline packages/core/src/boot/init.ts:52]: do we need to default to `[]` here?

File: packages/core/src/domain/configuration/experimentalFeatures.ts [hunk 6]
```diff
@@ -0,0 +1,26 @@
+let enabledExperimentalFeatures: Set<string>
+
+export function updateExperimentalFeatures(enabledFeatures: string[]): void {
```
Comment (bcaudan) [inline packages/core/src/domain/configuration/experimentalFeatures.ts:3]: since you have added extra code to handle some cases, what about adding some tests to document them and  avoid regression?

File: packages/core/src/boot/init.ts (hunk 5 above)
Comment (jagracey) [reply packages/core/src/boot/init.ts:52]: FMU,
Since the FF data is arriving externally from a JS snippet, we should have minimum validation to ensure we have an array since TS won't help runtime issues. To handle the case of being possibly undefined, we can either have this inside the TS file, or here. To me it makes sense to trust TS checking inside the func, so we should default to array outside the func.

Comment (bcaudan) [reply packages/core/src/boot/init.ts:52]: I though you were handling the case where no flags were provided here.

> To me it makes sense to trust TS checking inside the func

you are already doing some sanitization inside the function `Array.isArray` and `.filter((flag) => typeof flag === 'string')` which make me think you don't really trust the `enabledFeatures: string[]` typing.

what about doing all the sanitization at the same place?

File: packages/core/src/domain/configuration/experimentalFeatures.ts (hunk 6 above)
Comment (jagracey) [reply packages/core/src/domain/configuration/experimentalFeatures.ts:3]: just added some tests.

File: packages/core/src/boot/init.ts (hunk 5 above)
Comment (jagracey) [reply packages/core/src/boot/init.ts:52]: 1. enableExperimentalFeatures is typed within initConfig as 
`enableExperimentalFeatures?: string[] | undefined`
so we handle the undefined case outside the `updateExperimentalFeatures` with `initConfiguration.enableExperimentalFeatures || []`
2. Inside updateExperimentalFeatures, as an API, we IMO, shouldn't allow the undefined case in the context of typescript checking because we should be passing string[] only.
3. We can't depend on the runtime following the TS rules given customer control, so we handle them now inside `updateExperimentalFeatures` (changed from inside `commonInit` as a previous commit in this PR ).

Do you have an alternative in mind?

Comment (bcaudan) [reply packages/core/src/boot/init.ts:52]: > 1. enableExperimentalFeatures is typed within initConfig as
enableExperimentalFeatures?: string[] | undefined
so we handle the undefined case outside the updateExperimentalFeatures with initConfiguration.enableExperimentalFeatures || []
> 2. Inside updateExperimentalFeatures, as an API, we IMO, shouldn't allow the undefined case in the context of typescript checking because we should be passing string[] only.

why though?
do we have other use cases for `updateExperimentalFeatures` where we should not allow to pass `undefined`?

IMO, we could go with `updateExperimentalFeatures(initConfiguration.enableExperimentalFeatures)` or even `updateExperimentalFeatures(initConfiguration)` since we only use it from the `initConfiguration`.

it would allow us to handle the undefined case in the unit tests as well.

File: packages/core/src/domain/configuration/experimentalFeatures.ts [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+  // Safely handle external data
+  if (!Array.isArray(enabledFeatures)) {
+    return
+  }
+
+  if (!enabledExperimentalFeatures) {
+    enabledExperimentalFeatures = new Set(enabledFeatures)
+  }
```
Comment (bcaudan) [inline packages/core/src/domain/configuration/experimentalFeatures.ts:11]: no unit test for `!enabledExperimentalFeatures`, what was your intent?

File: packages/core/src/domain/configuration/experimentalFeatures.ts [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+
+  enabledFeatures
+    .filter((flag) => typeof flag === 'string')
+    .forEach((flag: string) => {
+      enabledExperimentalFeatures.add(flag)
+    })
+}
+
+export function isExperimentalFeatureEnabled(featureName: string): boolean {
+  return !!enabledExperimentalFeatures && enabledExperimentalFeatures.has(featureName)
```
Comment (bcaudan) [inline packages/core/src/domain/configuration/experimentalFeatures.ts:21]: no unit test for `!!enabledExperimentalFeatures`, what was your intent?

File: packages/core/src/domain/configuration/experimentalFeatures.ts [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+  if (!Array.isArray(enabledFeatures)) {
+    return
+  }
+
+  if (!enabledExperimentalFeatures) {
+    enabledExperimentalFeatures = new Set(enabledFeatures)
+  }
+
+  enabledFeatures
+    .filter((flag) => typeof flag === 'string')
```
Comment (bcaudan) [inline packages/core/src/domain/configuration/experimentalFeatures.ts:14]: no unit test for string filtering, do we need it?

File: packages/core/src/boot/init.ts (hunk 5 above)
Comment (amortemousque) [reply packages/core/src/boot/init.ts]: Would it be a valid solution to have a `sanitise` for the entire `initConfiguration`? This way we ensure to have a valid typescript object before any other logic.

Comment (jagracey) [reply packages/core/src/boot/init.ts]: Cool, switched to `updateExperimentalFeatures(initConfiguration.enableExperimentalFeatures)`

File: packages/core/src/domain/configuration/experimentalFeatures.spec.ts [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+  it('initial state is empty', () => {
+    resetExperimentalFeatures()
+    expect(isExperimentalFeatureEnabled(undefined)).toBeFalse()
+    expect(isExperimentalFeatureEnabled('foo')).toBeFalse()
+    expect(isExperimentalFeatureEnabled('bar')).toBeFalse()
+    expect(isExperimentalFeatureEnabled('')).toBeFalse()
+  })
+
+  it('updateExperimentalFeatures', () => {
+    updateExperimentalFeatures(['foo'])
```
Comment (bcaudan) [inline packages/core/src/domain/configuration/experimentalFeatures.spec.ts:17]: it would be preferable to run a `resetExperimentalFeatures` in an `afterEach` block in order to not leak features between tests

File: packages/core/src/domain/configuration/experimentalFeatures.spec.ts [hunk 11]
```diff
@@ -0,0 +1,39 @@
+import {
+  updateExperimentalFeatures,
+  isExperimentalFeatureEnabled,
+  resetExperimentalFeatures,
+} from './experimentalFeatures'
+
+describe('experimentalFeatures', () => {
+  it('initial state is empty', () => {
+    resetExperimentalFeatures()
+    expect(isExperimentalFeatureEnabled(undefined as any)).toBeFalse()
```
Comment (bcaudan) [inline packages/core/src/domain/configuration/experimentalFeatures.spec.ts:10]: do we want to use this API with `undefined`?

File: packages/core/src/domain/configuration/experimentalFeatures.spec.ts [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+    expect(isExperimentalFeatureEnabled('foo')).toBeFalse()
+    expect(isExperimentalFeatureEnabled('bar')).toBeFalse()
+    expect(isExperimentalFeatureEnabled('')).toBeFalse()
```
Comment (bcaudan) [inline packages/core/src/domain/configuration/experimentalFeatures.spec.ts:13]: do we want to use this API with `''`?

File: packages/core/src/domain/configuration/experimentalFeatures.spec.ts [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+  })
+
+  it('updateExperimentalFeatures', () => {
```
Comment (bcaudan) [inline packages/core/src/domain/configuration/experimentalFeatures.spec.ts:16]:
```suggestion
  it('should define enabled experimental features', () => {
```

File: packages/core/src/domain/configuration/experimentalFeatures.spec.ts [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+    updateExperimentalFeatures(['foo'])
+    expect(isExperimentalFeatureEnabled('foo')).toBeTrue()
+    expect(isExperimentalFeatureEnabled('bar')).toBeFalse()
+  })
+
+  it('resetExperimentalFeatures clears state', () => {
+    resetExperimentalFeatures()
+    expect(isExperimentalFeatureEnabled('foo')).toBeFalse()
+    expect(isExperimentalFeatureEnabled('bar')).toBeFalse()
+  })
```
Comment (bcaudan) [inline packages/core/src/domain/configuration/experimentalFeatures.spec.ts:26]: IMO since it is only meant to be used in unit tests, no need to add tests for it

File: packages/core/src/domain/configuration/experimentalFeatures.spec.ts [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+
+  it("functions won't throw", () => {
```
Comment (bcaudan) [inline packages/core/src/domain/configuration/experimentalFeatures.spec.ts:28]: what is your intent?

File: packages/core/src/domain/configuration/experimentalFeatures.ts (hunk 8 above)
Comment (jagracey) [reply packages/core/src/domain/configuration/experimentalFeatures.ts:21]: `!!enabledExperimentalFeatures` is tested when `isExperimentalFeatureEnabled` is called after a `resetExperimentalFeatures`.

File: packages/core/src/domain/configuration/experimentalFeatures.ts (hunk 7 above)
Comment (jagracey) [reply packages/core/src/domain/configuration/experimentalFeatures.ts:11]: You cannot add flags to a set if the set does not exist. This logic is tested via lines 37,39,40.

File: packages/core/src/domain/configuration/experimentalFeatures.spec.ts (hunk 12 above)
Comment (jagracey) [reply packages/core/src/domain/configuration/experimentalFeatures.spec.ts:13]: Just checking as a valid common string. I haven't seen a spec around which string types we care about (uppercase, "Turkish dotless i", diacritics, emoji). 

Do we have a team preference on the extent of string validation?

File: packages/core/src/domain/configuration/experimentalFeatures.spec.ts (hunk 15 above)
Comment (jagracey) [reply packages/core/src/domain/configuration/experimentalFeatures.spec.ts:28]: Testing via unusual calling patterns given sometimes errors can be thrown with unexpected duplicate calls.
Do we have a team preference here on test scope?

Comment (jagracey) [reply packages/core/src/domain/configuration/experimentalFeatures.spec.ts]: Updated, moved the string filtering test here.

File: packages/core/src/domain/configuration/experimentalFeatures.spec.ts [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+    resetExperimentalFeatures()
+    resetExperimentalFeatures()
+    updateExperimentalFeatures(undefined)
+    updateExperimentalFeatures(undefined)
+    expect(isExperimentalFeatureEnabled('foo')).toBeFalse()
+
+    updateExperimentalFeatures(['foo'])
+    updateExperimentalFeatures([11 as any, [] as any, {} as any, null as any, undefined as any])
+    expect(isExperimentalFeatureEnabled('foo')).toBeTrue()
+  })
```
Comment (bcaudan) [inline packages/core/src/domain/configuration/experimentalFeatures.spec.ts:33]:
```suggestion
  it('should allow to be shared between products', () => {
    updateExperimentalFeatures(['foo'])
    updateExperimentalFeatures(['bar'])
    
    expect(isExperimentalFeatureEnabled('foo')).toBeTrue()
    expect(isExperimentalFeatureEnabled('bar')).toBeTrue()
  })
  
  it('should support some edge cases', () => {
    updateExperimentalFeatures(['foo'])
    updateExperimentalFeatures(undefined)
    updateExperimentalFeatures([])
    updateExperimentalFeatures([11 as any])
    
    expect(isExperimentalFeatureEnabled('foo')).toBeTrue()
  })
```

**Review by bcaudan — APPROVED** [review]
